tests/contrib/gunicorn/test_gunicorn.py

```diff
@@ -52,79 +94,103 @@ def _gunicorn_settings_factory(
         worker_class=worker_class,
         bind=bind,
         use_ddtracerun=use_ddtracerun,
-        post_worker_init=post_worker_init,
+        import_sitecustomize_in_postworkerinit=import_sitecustomize_in_postworkerinit,
+        start_service_in_hook_named=start_service_in_hook_named,
     )
 
 
-@pytest.fixture
-def gunicorn_server_settings():
-    yield _gunicorn_settings_factory()
-
-
-@pytest.fixture
-def gunicorn_server(gunicorn_server_settings, tmp_path):
-    cfg_file = tmp_path / "gunicorn.conf.py"
+def build_config_file(gunicorn_server_settings):
+    post_fork = START_SERVICE if gunicorn_server_settings.start_service_in_hook_named != "post_worker_init" else ""
+    post_worker_init = "    {sitecustomize}\n{service_start}".format(
+        sitecustomize=IMPORT_SITECUSTOMIZE if gunicorn_server_settings.import_sitecustomize_in_postworkerinit else "",
+        service_start=START_SERVICE
+        if gunicorn_server_settings.start_service_in_hook_named == "post_worker_init"
+        else "",
+    )
     cfg = """
+def post_fork(server, worker):
+    pass
+{post_fork}
+
+def post_worker_init(worker):
+    pass
 {post_worker_init}
 
 workers = {num_workers}
 worker_class = "{worker_class}"
 bind = "{bind}"
 """.format(
-        post_worker_init=gunicorn_server_settings.post_worker_init,
+        post_fork=post_fork,
+        post_worker_init=post_worker_init,
         bind=gunicorn_server_settings.bind,
         num_workers=gunicorn_server_settings.num_workers,
         worker_class=gunicorn_server_settings.worker_class,
     )
+    return cfg
+
+
+@contextmanager
+def gunicorn_server(gunicorn_server_settings, tmp_path):
+    cfg_file = tmp_path / "gunicorn.conf.py"
+    cfg = build_config_file(gunicorn_server_settings)
     cfg_file.write_text(stringify(cfg))
     cmd = []
     if gunicorn_server_settings.use_ddtracerun:
         cmd = ["ddtrace-run"]
     cmd += ["gunicorn", "--config", str(cfg_file), str(gunicorn_server_settings.app_path)]
     print("Running %r with configuration file %s" % (" ".join(cmd), cfg))
-    p = subprocess.Popen(
+
+    server_process = subprocess.Popen(
         cmd,
         env=gunicorn_server_settings.env,
         cwd=gunicorn_server_settings.directory,
-        stdout=sys.stdout,
-        stderr=sys.stderr,
+        stderr=subprocess.PIPE,
         close_fds=True,
         preexec_fn=os.setsid,
     )
     try:
         client = Client("http://%s" % gunicorn_server_settings.bind)
-        # Wait for the server to start up
         try:
-            client.wait(max_tries=20, delay=0.5)
+            client.wait(max_tries=100, delay=0.1)
         except tenacity.RetryError:
-            # if proc.returncode is not None:
-            # process failed
             raise TimeoutError("Server failed to start, see stdout and stderr logs")
-
-        yield client
-
+        time.sleep(SERVICE_INTERVAL)
```

Comment:
Out of curiousity, what does waiting here before yielding accomplish?

Reply:
It gives the `PeriodicService` time to wake up and try to run. It's only when that happens that the service's zombie threads self-destruct. This is definitely worth adding a comment about.

Reply:
I see, a comment here definitely sounds helpful!